### ISSUE TITLE
Provisioned Concurrency without DeploymentPreference causes failure on second deployment 

### ISSUE THREAD

**Issue by jcts3** [aws/serverless-application-model#1296]
**Description:**

Using new Provisioned Concurrency capabilities without a `DeploymentPreference` Property causes a failed update (and consequent failed update rollback) on the second deployment. 

I don't think it's an issue with SAM persay - just that it needs to be checked for, and the documentation needs to be updated, i.e. the Release note for v1.18.0 will cause a failed update (and failed rollback) on second deployment. Potentially also add a comment on this table in the bottom 2 rows: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#properties ?


**Steps to reproduce the issue:**
1. Deploy a lambda with appropriate `ProvisionedConcurrencyConfig` property, but without a `DeploymentPreference` property. This deploys succesfully and works as expected. 
2. Make any change to the lambda (e.g. just add a '!' to the hello world message)
3. Build and deploy again.

I reproduced this using this hello world project: https://github.com/jcts3/sam-pc-experiment/blob/master/template.yaml 

**Observed result:**
Function Alias fails to update. 
`Alias with weights can not be used with Provisioned Concurrency (Service: AWSLambdaInternal; Status Code: 400; Error Code: InvalidParameterValueException; Request ID: 15023437-6d75-40d3-ae4c-313691fcf174)`

`Invalid function version 1. Function version 1 is already included in                                                                                                                                                                                                          routing configuration. (Service: AWSLambdaInternal; Status Code: 400; Error Code: InvalidParameterValueException; Request ID: 92ee7215-037c-417f-8874-0f024bc8022f)`

I think what's happening is the alias is trying to shift from version 1to version 2 of the function, but the provisioned concurrency isn't allowing it to do this by itself (i.e. without CodeDeploy managing the change).

**Expected result:**
Update to not fail. Alias to switch between versions. 
OR 
First deployment to not succeed due to it missing a required property

**Proposed Fix**
Update docs to reflect requirement of `DeploymentPreference` property when using `ProvisionedConcurrencyConfig` property. 
Update pre-deployment checks to prevent deployment without `DeploymentPreference` when using `ProvisionedConcurrencyConfig` property.

**Comment by praneetap** [comment]
@jcts3 thanks for reporting this issue! This is such a helpful issue description. I could reproduce it on my end. I am reaching out to the right people, meanwhile, I agree with you on updating the docs. I don't think we should add a pre-deployment check yet, but let me investigate further and get back to you.

**Comment by GrahamMThomas** [comment]
I was having this same problem and here is a **workaround** that worked for me until this is fixed:

```
      AutoPublishAlias: live
      ProvisionedConcurrencyConfig:
        ProvisionedConcurrentExecutions: 1
      DeploymentPreference:
        Type: AllAtOnce
```

**Comment by zd987** [comment]
I notice another issue: if you have not deployed ProvisionedConcurrencyConfig before, and the first time you introduce ProvisionedConcurrencyConfig together with "DeploymentPreference: Type: AllAtOnce", the ProvisionedConcurrencyConfig is not deployed successfully. 

So it looks like you need to first introduce ProvisionedConcurrencyConfig without DeploymentPreference, and later if you want to modify this lambda, then add DeploymentPreference as mentioned in the previous posts here.

**Comment by aaronwht** [comment]
Unable to run SAM (v 0.34.0) locally using `sam local start-api --port 3030 --skip-pull-image` when including the below in template file.  Works in production and not locally.
AutoPublishAlias: live
      ProvisionedConcurrencyConfig:
        ProvisionedConcurrentExecutions: 5
      DeploymentPreference:
        Type: AllAtOnce

Error: [InvalidResourceException('Index', 'property ProvisionedConcurrencyConfig not defined for resource of type AWS::Serverless::Function')] ('Index', 'property ProvisionedConcurrencyConfig not defined for resource of type AWS::Serverless::Function')

**Comment by praneetap** [comment]
> Unable to run SAM (v 0.34.0) locally using `sam local start-api --port 3030 --skip-pull-image` when including the below in template file. Works in production and not locally.
> 
> ```
> AutoPublishAlias: live
>       ProvisionedConcurrencyConfig:
>         ProvisionedConcurrentExecutions: 5
>       DeploymentPreference:
>         Type: AllAtOnce
> ```
> 
> Error: [InvalidResourceException('Index', 'property ProvisionedConcurrencyConfig not defined for resource of type AWS::Serverless::Function')] ('Index', 'property ProvisionedConcurrencyConfig not defined for resource of type AWS::Serverless::Function')

Can you update your sam-cli to the latest version?

**Comment by sriram-mv** [comment]
Is this still an issue with latest version of SAM CLI? Closing this issue, but will re-open if its not solved.

**Comment by guaccc** [comment]
I broke my codepipeline deployment after trying to add provisioned concurrency per docs. It worked at first but the next time I ran my pipeline it broke with same error `Alias with weights can not be used with Provisioned Concurrency`.

Now I cannot seem to back out of it! I've tried removing the parameters from cloudformation and rerunning, and manually removing both the parameters, the provisioned concurrency, and the alias from the functions simultaneously in a new run and it's still happening.

Any ideas how to break out of this without nuking my entire stack and recreating the functions?

**Comment by ujjwal3067** [comment]
still experiencing this issue. Can someone confirm if this still an issue ?

**Comment by priyakhandelwal** [comment]
Facing the same issue.

**Comment by debojit007** [comment]
### Request
Please guide on how to use  `alias` and `provisionedConcurrency` on my `SecureFunction `lambda without running into CFN deployment failure

### Setup Background :
- My CDK code uses `SecureFunction` lambda that *initially* had `alias` with `provisionedConcurrency` set to 100 in Gamma stage of my pipeline
- Gamma stage was only in present in IAD
- After 1st deployment lambda was working fine

### Issue 
- Due to project need I made changes in CDK to :
  - set `provisionedConcurrency` as *1* in Gamma
  - expand Gamma to PDX & DUB regions
  - **NO CHANGES** in Lambda code
- `bootstrap` command for region expansion to take effect in pipeline & checked in code to pipeline
- Deployment repeatedly failed in Beta stage with misleading error `Provisioned Concurrency configuration failed to be applied. Reason: FUNCTION_ERROR_INIT_FAILURE`
- Deleted stack and tried redeploying but then started seeing another strange error `Alias with weights can not be used with Provisioned Concurrency`
- Beta stage of pipeline blocked and deployment not successful

### Intermediary fix

- I removed use of `alias and provisionedConcurrency` in my CDK code
- Pushed code change to pipeline
- Verified that lambda behaviour is working fine.
- Beta stage unblocked with successful deployment
- Making it clear that issue is with the changes on `alias` and `provisionedConcurrency` on subsequent deployments involving `SecureFunction` lambda

**Comment by xazhao** [comment]
Trying to reproduce using steps provided:
```
1. Deploy a lambda with appropriate ProvisionedConcurrencyConfig property, but without a DeploymentPreference property. This deploys succesfully and works as expected.
2. Make any change to the lambda (e.g. just add a '!' to the hello world message)
3. Build and deploy again.
```
I can build/deploy without any issues. Seems like it has been fixed.

@debojit007 You question looks like a question to CDK. You could ask in their Github repo: https://github.com/aws/aws-cdk

Closing the issue for now. Free to re-open it if the issue persists.
